sync/syncer.go

```diff
@@ -145,97 +148,175 @@ func (s *Syncer) Synchronise() {
+			continue
+		}
+		s.Info("add block to layer %v", block)
+		if err := s.AddBlock(block); err != nil {
+			s.Warning(fmt.Sprintf("could not add %v", block.ID()), err)
+			continue
+		}
+		s.Info("added block to layer %v", block)
 		blocksArr = append(blocksArr, block)
 
 	}
 
-	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
-}
-
-type peerHashPair struct {
-	peer p2p.Peer
-	hash []byte
+	return blocksArr, nil
 }
 
-func (s *Syncer) getLayerBlockIDs(index types.LayerID) (chan types.BlockID, error) {
-
-	m, err := s.fetchLayerHashes(index)
+func (s *Syncer) Txs(mb *types.MiniBlock) ([]*types.SerializableTransaction, error) {
+	foundTxs, missing := s.GetTransactions(mb.TxIds)
+	//map and sort txs
+	txMap := make(map[types.TransactionId]*types.SerializableTransaction)
+	if len(missing) > 0 {
+		for out := range s.fetchWithFactory(TxReqFactory(missing), 1) {
+			ntxs := out.([]types.SerializableTransaction)
+			for _, tx := range ntxs {
+				txMap[types.GetTransactionId(&tx)] = &tx
+			}
+		}
+	}
 
-	if err != nil {
-		s.Error("could not get LayerHashes for layer: %v", index)
-		return nil, err
+	txs := make([]*types.SerializableTransaction, 0, len(mb.TxIds))
+	for _, t := range mb.TxIds {
+		if tx, ok := foundTxs[t]; ok {
+			txs = append(txs, tx)
+		} else if tx, ok := txMap[t]; ok {
+			txs = append(txs, tx)
+		} else {
+			return nil, errors.New(fmt.Sprintf("could not fetch tx %v", t))
+		}
 	}
-	return s.fetchLayerBlockIds(m, index)
+	return txs, nil
 }
 
-func keysAsChan(idSet map[types.BlockID]struct{}) chan types.BlockID {
-	res := make(chan types.BlockID, len(idSet))
-	defer close(res)
-	for id := range idSet {
-		res <- id
+func (s *Syncer) ATXs(mb *types.MiniBlock) ([]*types.ActivationTx, error) {
+	foundTxs, missing := s.GetATXs(mb.ATxIds)
+	//map and sort txs
+	txMap := make(map[types.AtxId]*types.ActivationTx)
+	if len(missing) > 0 {
+		output := s.fetchWithFactory(ATxReqFactory(missing), 1)
+		for out := range output {
+			ntxs := out.([]types.ActivationTx)
+			for _, atx := range ntxs {
+				txMap[atx.Id()] = &atx
+			}
+		}
+	}
+
+	atxs := make([]*types.ActivationTx, 0, len(mb.TxIds))
+	for _, t := range mb.ATxIds {
+		if tx, ok := foundTxs[t]; ok {
+			atxs = append(atxs, tx)
+		} else if tx, ok := txMap[t]; ok {
+			atxs = append(atxs, tx)
+		} else {
+			return nil, errors.New(fmt.Sprintf("could not fetch atx %v", t))
+		}
 	}
-	return res
+	return atxs, nil
 }
 
-func (s *Syncer) fetchLayerBlockIds(m map[string]p2p.Peer, lyr types.LayerID) (chan types.BlockID, error) {
-	// each worker goroutine tries to fetch a block iteratively from each peer
-	peers := s.GetPeers()
-	if len(peers) == 0 {
-		return nil, errors.New("no peers")
+func (s *Syncer) fetchLayerBlockIds(m map[string]p2p.Peer, lyr types.LayerID) ([]types.BlockID, error) {
+	//send request to different users according to returned hashes
+	v := make([]p2p.Peer, 0, len(m))
+	for _, value := range m {
```

Comment:
you are right i guess i just didn't see a reason to validate
because this is basically an optimisation.
 we are trying to save on requests to peers with the same information or else we would just ask everybody for the ids they know.

Reply:
I understand, validating the hash is indeed not necessary. 
There are several optimizations that we can do here, for example, asking the neighbors for a list of block hashes and fetch each block once. I'll open a ticket on some stuff that I have in mind.